hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/run/strategy/MultipleSparkJobExecutionStrategy.java

```diff
@@ -95,8 +95,7 @@ public MultipleSparkJobExecutionStrategy(HoodieTable table, HoodieEngineContext
         .map(inputGroup -> runClusteringForGroupAsync(inputGroup,
             clusteringPlan.getStrategy().getStrategyParams(),
             Option.ofNullable(clusteringPlan.getPreserveHoodieMetadata()).orElse(false),
-            instantTime))
-        .map(CompletableFuture::join);
+            instantTime)).collect(Collectors.toList()).stream().map(CompletableFuture::join);
```

Comment:
How does this guarantee jobs will run in parallel? We simply dereference stream into list, but then still join the Futures sequentially.

Instead we should use following util 

```
public static <T> CompletableFuture<List<T>> allOf(@Nonnull List<CompletableFuture<T>> futures) {
        return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
            .thenApply(aVoid ->
                futures.stream()
                    // NOTE: This join wouldn't block, since all the
                    //       futures are completed at this point
                    .map(CompletableFuture::join)
                    .collect(Collectors.toList())
            );
    }
```

And then invoke it like following 

```
allOf(
  clusteringPlan.getInputGroups()
    .stream()
    .map(...) // returns `CompletableFuture`
    .collect(Collectors.toList())
)
  .join();
```

This would guarantee parallel execution for each individual clustering group

Reply:
@alexeykudinkin   thanks for your review.  
no,  this change can guarantee jobs will run in parallel. we should do collect(Collectors.toList()) before join the future.
the test result and our user test result(see [SUPPORT] Zordering clustering on a moderate size dataset taking large amounts of time. #4135 )  see this modify is ok.

Reply:
That's exactly my question: why do you think this will guarantee that?

Reply:
@alexeykudinkin，  here is my explain：
1） in original code， one streaming pipleline is used， using join in the same pipeline will lead threads to execute one by one
2） Here are two stream operations. Join operation is in the second stream pipleline， this operation will not lead  threads to execute one by one。

public static void main(String[] args) {
    Integer[] test = new Integer[] {0, 2, 3};
    long time = System.currentTimeMillis();
    List<CompletableFuture<Integer>> ls =
        Arrays.stream(test).map(f -> CompletableFuture.supplyAsync(() -> {
        System.out.println(Thread.currentThread().getName());
        try {
            Thread.sleep(10000);
        } catch (InterruptedException e) {
        }
        return f;
    })).collect(Collectors.toList());
    ls.stream().map(f  -> f.join()).collect(Collectors.toList());
    System.out.println(String.format("cost time: %s", (System.currentTimeMillis() - time) /1000));
}





You are an expert in this field， thank you for your patient guidance. If you still think there is a problem, I will modify the code. Thanks again

Reply:
@xiarixiaoyao I see where we diverge now: you're relying on stream parallelism which is very implicit behavior, and as you might have noticed prior stream parallelism is not guaranteed (some operations in the current JDK impl are actually implicitly converting the stream into "sequential" one from parallel w/o you even knowing it). 

There's also another angle, that joining Futures w/in Streams will be blocking threads of `ForkJoinPool.common` executor.

Hence is my proposal to instead rely on _explicit_ conversion using `allOf` and subsequent join.

Reply:
got it， thanks， will update the code